### PR TITLE
feat(db): Add mongo, User and Post models

### PR DESCRIPTION
**Pull request by shaine**

### CODE REVIEW

File: src/server.js [hunk 1]
```diff
@@ -103,9 +107,11 @@ app.use((req, res) => renderError(req, res, 404, 'Not Found'));
 app.use((error, req, res, next) => // eslint-disable-line
     console.error(error) || renderError(req, res, 500, 'Internal Server Error'));
 
-server.listen(port, err => {
-    if (err) {
-        console.error(err, true);
-    }
-    console.info(`----\n==> ✅  Server is running on port ${port}.`);
+dbConnect(() => {
```
Comment (shaine) [inline src/server.js]: <3 CPS

File: .gitignore [hunk 2]
```diff
@@ -7,6 +7,7 @@ public/*
 private/*
 content/*
 TODO.md
+db
```
Comment (shaine) [inline .gitignore]: Storing the DB in the actual project. Dealing with many projects spread across many systems/directories in the past has been annoying.

File: karma.conf.js [hunk 3]
```diff
@@ -1,115 +0,0 @@
-const webpack = require('webpack');
```
Comment (shaine) [inline karma.conf.js]: Bye bitch

File: scripts/mocha.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env node_modules/.bin/babel-node
```
Comment (shaine) [inline scripts/mocha.js]: Just testing with mocha <3

File: src/helpers/db.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
+const mongoose = require('mongoose');
```
Comment (shaine) [inline src/helpers/db.js]: Just testing with mocha means no imports/exports. I guess I could use babel as a mocha transform but meh, maybe later

File: src/helpers/db.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+const {
+    dbUri
+} = require('../../config.json');
+
+module.exports = function connect(cb) {
+    mongoose.connect(dbUri);
+    const db = mongoose.connection;
```
Comment (shaine) [inline src/helpers/db.js]: Mongoose is a great mongo adapter. Very easy to use.

File: src/helpers/db.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+    db.on('error', error => {
+        console.error('DB connection error', error);
+        cb(error);
+    });
+
+    db.once('open', () => {
+        console.info('----\n==> ✅  Connected to DB');
+        cb(null, db);
+    });
```
Comment (shaine) [inline src/helpers/db.js]: This file is separated out like this so I can use it in the repl. It doesn't really do anything interesting on its own.

File: src/helpers/password.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
+const sodium = require('sodium').api;
```
Comment (shaine) [inline src/helpers/password.js]: Okay so sodium is way cool. It yields an argon 2 lib, which is I guess the successor to bcrypt for password hashing. It has a bunch of other handy encryption functions within it as well.

File: src/helpers/password.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+exports.hashPassword = function hashPassword(password) {
```
Comment (shaine) [inline src/helpers/password.js]: Pretty straight forward. Password goes in, hash comes out. The hash contains the salt, and the variety of other details that the algo will need to verify subsequent passwords - userland doesn't have to worry about storing them separately.

File: src/helpers/password.js [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+    const hashBuffer = sodium.crypto_pwhash_str(
+        passwordBuffer,
+        sodium.crypto_pwhash_OPSLIMIT_INTERACTIVE,
+        sodium.crypto_pwhash_MEMLIMIT_INTERACTIVE
+    );
+
+    return hashBuffer.toString();
+}
+
+exports.verifyPassword = function verifyPassword(password, passwordHash) {
```
Comment (shaine) [inline src/helpers/password.js]: Simple verification.

File: src/models/Post.js [hunk 11]
```diff
@@ -0,0 +1,21 @@
+const mongoose = require('mongoose');
+
+const postSchema = mongoose.Schema({
+    title: String,
+    content: String,
+    authorId: mongoose.Schema.Types.ObjectId,
+    createdAt: { type: Date, default: Date.now }
+});
+
+const Post = mongoose.model('Post', postSchema);
```
Comment (shaine) [inline src/models/Post.js]: Mongoose defines a schema, then creates a model off of it. Not presently certain why they are separate entities.

File: src/models/Post.test.js [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+const { expect } = chai;
+chai.use(sinonChai);
+
+const PostStub = sinon.stub();
+PostStub.findOne = sinon.stub();
+PostStub.prototype.save = sinon.stub();
+const mongooseStub = {
+    Schema: sinon.stub().returns('foo schema'),
+    model: sinon.stub().withArgs('Post', 'foo schema').returns(PostStub)
+};
```
Comment (shaine) [inline src/models/Post.test.js]: The fact that I'm instantiating and saving the model all in one function means that my unit tests have to do really screwy things to hack into the implementation to verify. Not stellar. This is the unit test telling me that I'm designing these units wrong, a warning which I am going to ignore.

File: src/models/User.js [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+const userSchema = mongoose.Schema({
+    name: String,
+    email: String,
+    passwordHash: String,
+    createdAt: { type: Date, default: Date.now }
+});
+
+const User = mongoose.model('User', userSchema);
+
+exports.getUserByEmailAndPassword = function getUserByEmailAndPassword(email, password, cb) {
```
Comment (shaine) [inline src/models/User.js]: One of the things (some) people learned in Rails is that you shouldn't expose your damn ORM throughout your entire application. Encapsulate your models and expose only the concerns your application is concerned with. There's no reason for your views to have access to advanced ORM features.

File: src/models/User.js [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    const passwordHash = hashPassword(password);
+
+    const userInfo = {
+        name,
+        email,
+        passwordHash
+    };
+
+    const user = new User(userInfo);
+    user.save(cb);
```
Comment (shaine) [inline src/models/User.js]: I should probably update this to not return the mongoose model - ideally, the model never escapes this file, but instead yields a primitive application-friendly copy of it.

File: src/models/User.test.js [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+            }, cbStub);
+
+            expect(UserStub.prototype.save)
+                .to.have.been.calledWith(cbStub);
+            expect(UserStub)
+                .to.have.been.calledWith({
+                    name: 'foo',
+                    email: 'bar@test.com',
+                    passwordHash: 'hash 123'
+                });
```
Comment (shaine) [inline src/models/User.test.js]: These expectations really aren't that effective against the SUT. You can tell this part wasn't TDD.

File: karma.conf.js (hunk 3 above)
Comment (elliotec) [inline karma.conf.js]: yeah bitch

**Pull request merged**